File: tunnel-client/src/test/java/com/tuuzed/tunnel/client/TunnelClientTest.java
package com.tuuzed.tunnel.client;

import com.tuuzed.tunnel.common.log4j.Log4jInitializer;
import com.tuuzed.tunnel.common.proto.ProtoRequest;
import com.tuuzed.tunnel.common.util.HttpUtils;
import com.tuuzed.tunnel.common.util.SslContexts;
import io.netty.handler.ssl.SslContext;
import org.apache.log4j.Level;
import org.jetbrains.annotations.NotNull;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TunnelClientTest {
    private TunnelClient client;
    private ProtoRequest portError;
    private ProtoRequest tcpHttp;
    private ProtoRequest vnc;
    private ProtoRequest ssh;
    private ProtoRequest vhostHttp1;
    private ProtoRequest vhostHttp2;
    private ProtoRequest vhostHttps1;
    private ProtoRequest vhostHttps2;
    private ProtoRequest portReplaced;


    @Test
    public void start() throws Exception {
        final String serverAddr = "127.0.0.1";
        final int serverPort = 5001;
        final SslContext sslContext = SslContexts.forClient("../resources/jks/client.jks", "ctunnelpass");

        client.connect(serverAddr, serverPort, portError, sslContext);

        client.connect(serverAddr, serverPort, tcpHttp, sslContext);
        client.connect(serverAddr, serverPort, vnc, sslContext);
        client.connect(serverAddr, serverPort, ssh, sslContext);
        // vhostHttp
        client.connect(serverAddr, serverPort, vhostHttp1, sslContext);
        client.connect(serverAddr, serverPort, vhostHttp2, sslContext);
        // vhostHttps
        client.connect(serverAddr, serverPort, vhostHttps1, sslContext);
        client.connect(serverAddr, serverPort, vhostHttps2, sslContext);

        final TunnelClientDescriptor portReplacedDescriptor = client.connect(serverAddr, serverPort, portReplaced, sslContext);
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    Thread.sleep(6_000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                portReplacedDescriptor.shutdown();
            }
        }).start();
        Thread.currentThread().join();
    }

    @Before
    public void setUp() {
        Log4jInitializer.initializeThirdLibrary(Level.WARN);
        Log4jInitializer.builder().initialize();

        client = TunnelClient.builder()
            .setAutoReconnect(true)
            .setWorkerThreads(4)
            .setListener(new TunnelClientListener() {
                private final Logger logger = LoggerFactory.getLogger(TunnelClientListener.class);

                @Override
                public void onConnecting(@NotNull TunnelClientDescriptor descriptor, boolean reconnect) {
                    logger.warn("tunnel: {}, reconnect: {}", descriptor, reconnect);
                }

                @Override
                public void onConnected(@NotNull TunnelClientDescriptor descriptor) {
                    logger.error("{}", descriptor);
                }

                @Override
                public void onDisconnect(@NotNull TunnelClientDescriptor descriptor, boolean fatal) {
                    logger.error("tunnel: {}, deadly: {}", descriptor, fatal);
                }
            })
            .build();

        portError = ProtoRequest.tcpBuilder(65000)
            .setLocalAddr("192.168.1.1")
            .setLocalPort(80)
            .setToken("tk123456")
            .build();

        portReplaced = ProtoRequest.tcpBuilder(20000) // 20080
            .setLocalAddr("192.168.1.1")
            .setLocalPort(80)
            .setToken("tk123456")
            .build();

        tcpHttp = ProtoRequest.tcpBuilder(10080) // 20080
            .setLocalAddr("192.168.1.1")
            .setLocalPort(80)
            .setOption("token", "tk123456")
            .build();

        vnc = ProtoRequest.tcpBuilder(15900) // 20080
            .setLocalAddr("192.168.1.33")
            .setLocalPort(5900)
            .setToken("tk123456")
            .build();

        ssh = ProtoRequest.tcpBuilder(10022) // 20080
            .setLocalAddr("192.168.1.10")
            .setLocalPort(22)
            .setToken("tk123456")
            .build();

        vhostHttp1 = ProtoRequest.httpBuilder("t1.tunnel.lo")
            .setLocalAddr("192.168.1.1")
            .setLocalPort(80)
            .setToken("tk123456")
            .setRewriteHeaders(HttpUtils.headersOf("X-Real-IP", "$remote_addr"))
            .setWriteHeaders(HttpUtils.headersOf("X-User-Agent", "Tunnel"))
            .build();

        vhostHttp2 = ProtoRequest.httpBuilder("t2.tunnel.lo")
            .setLocalAddr("111.230.198.37")
            .setLocalPort(10080)
            .setToken("tk123456")
            .setRewriteHeaders(HttpUtils.headersOf("X-Real-IP", "$remote_addr"))
            .setWriteHeaders(HttpUtils.headersOf("X-User-Agent", "Tunnel"))
            .setBasicAuth(true, "OCR")
            .setBasicAuthAccount("admin", "admin")
            .build();


        vhostHttps1 = ProtoRequest.httpsBuilder("t1.tunnel.lo")
            .setLocalAddr("192.168.1.1")
            .setLocalPort(80)
            .setToken("tk123456")
            .setRewriteHeaders(HttpUtils.headersOf("X-Real-IP", "$remote_addr"))
            .setWriteHeaders(HttpUtils.headersOf("X-User-Agent", "Tunnel"))
            .build();

        vhostHttps2 = ProtoRequest.httpsBuilder("t2.tunnel.lo")
            .setLocalAddr("111.230.198.37")
            .setLocalPort(10080)
            .setToken("tk123456")
            .setRewriteHeaders(HttpUtils.headersOf("X-Real-IP", "$remote_addr"))
            .setWriteHeaders(HttpUtils.headersOf("X-User-Agent", "Tunnel"))
            .setBasicAuth(true, "OCR")
            .setBasicAuthAccount("admin", "admin")
            .build();
    }

    @After
    public void shutDown() {
        client.destroy();
    }


}

File: tunnel2-client/src/test/kotlin/tunnel2/client/TunnelClientTest.kt
package tunnel2.client

import org.apache.log4j.Level
import org.junit.After
import org.junit.Before
import org.junit.Test
import tunnel2.common.TunnelRequest
import tunnel2.common.logging.LoggerFactory
import tunnel2.common.ssl.SslContexts

class TunnelClientTest {
    private lateinit var client: TunnelClient
    private lateinit var portError: TunnelRequest
    private lateinit var tcpHttp: TunnelRequest
    private lateinit var vnc: TunnelRequest
    private lateinit var ssh: TunnelRequest
    private lateinit var vhostHttp1: TunnelRequest
    private lateinit var vhostHttp2: TunnelRequest
    private lateinit var vhostHttps1: TunnelRequest
    private lateinit var vhostHttps2: TunnelRequest
    private lateinit var portReplaced: TunnelRequest


    @Test
    @Throws(Exception::class)
    fun start() {
        val serverAddr = "127.0.0.1"
        val serverPort = 5001
        val sslContext = SslContexts.forClient(
            "../resources/jks/t2c.jks",
            "t2cpass"
        )
        client.connect(serverAddr, serverPort, portError, sslContext)

        client.connect(serverAddr, serverPort, tcpHttp, sslContext)
        client.connect(serverAddr, serverPort, vnc, sslContext)
        client.connect(serverAddr, serverPort, ssh, sslContext)
        // vhostHttp
        client.connect(serverAddr, serverPort, vhostHttp1, sslContext)
        client.connect(serverAddr, serverPort, vhostHttp2, sslContext)
        // vhostHttps
        client.connect(serverAddr, serverPort, vhostHttps1, sslContext)
        client.connect(serverAddr, serverPort, vhostHttps2, sslContext)

        val portReplacedDescriptor = client.connect(serverAddr, serverPort, portReplaced, sslContext)
        Thread(Runnable {
            try {
                Thread.sleep(6000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

            portReplacedDescriptor.shutdown()
        }).start()
        Thread.currentThread().join()
    }

    @Before
    fun setUp() {
        LoggerFactory.configConsole(level = Level.WARN, names = *LoggerFactory.thirdLibs).apply()
        LoggerFactory.configConsole(level = Level.ALL).apply()
        LoggerFactory.configFile(level = Level.ALL, file = "../logs/t2c.log").apply()

        client = TunnelClient(
            autoReconnect = true,
            workerThreads = 4,
            listener = object : TunnelClient.Listener {
                private val logger = LoggerFactory.getLogger(TunnelClient.Listener::class.java)

                override fun onConnecting(descriptor: TunnelClientDescriptor, reconnect: Boolean) {
                    logger.warn("tunnel: {}, reconnect: {}", descriptor, reconnect)
                }

                override fun onConnected(descriptor: TunnelClientDescriptor) {
                    logger.error("{}", descriptor)
                }

                override fun onDisconnect(descriptor: TunnelClientDescriptor, err: Boolean, errCause: Throwable?) {
                    logger.error("tunnel: {}, err: {}", descriptor, err, errCause)
                }
            }
        )

        portError = TunnelRequest.ofTcp(
            "192.168.1.1",
            80,
            65000,
            "tk123456"
        )
        portReplaced = TunnelRequest.ofTcp(
            "192.168.1.1",
            80,
            20000,
            "tk123456"
        ) //  replaced 20080

        tcpHttp = TunnelRequest.ofTcp(
            "111.230.198.37",
            10080,
            10080,
            "tk123456"
        )

        vnc = TunnelRequest.ofTcp(
            "192.168.1.33",
            5900,
            15900,
            "tk123456"
        )
        ssh = TunnelRequest.ofTcp(
            "192.168.1.10",
            22,
            10022,
            "tk123456"
        )

        vhostHttp1 = TunnelRequest.ofHttp(
            localAddr = "192.168.1.1",
            localPort = 80,
            vhost = "t1.tunnel.lo",
            authToken = "tk123456",
            proxySetHeaders = mapOf(Pair("X-Real-IP", "\$remote_addr")),
            proxyAddHeaders = mapOf(Pair("X-User-Agent", "Tunnel"))
        )

        vhostHttp2 = TunnelRequest.ofHttp(
            localAddr = "111.230.198.37",
            localPort = 10080,
            vhost = "t2.tunnel.lo",
            enableBasicAuth = true,
            basicAuthRealm = "OCR",
            basicAuthUsername = "admin",
            basicAuthPassword = "admin",
            authToken = "tk123456",
            proxySetHeaders = mapOf(Pair("X-Real-IP", "\$remote_addr")),
            proxyAddHeaders = mapOf(Pair("X-User-Agent", "Tunnel"))
        )

        vhostHttps1 = TunnelRequest.ofHttps(
            localAddr = "192.168.1.1",
            localPort = 80,
            vhost = "t1.tunnel.lo",
            authToken = "tk123456",
            proxySetHeaders = mapOf(Pair("X-Real-IP", "\$remote_addr")),
            proxyAddHeaders = mapOf(Pair("X-User-Agent", "Tunnel"))
        )

        vhostHttps2 = TunnelRequest.ofHttps(
            localAddr = "111.230.198.37",
            localPort = 10080,
            vhost = "t2.tunnel.lo",
            enableBasicAuth = true,
            basicAuthRealm = "OCR",
            basicAuthUsername = "admin",
            basicAuthPassword = "admin",
            authToken = "tk123456",
            proxySetHeaders = mapOf(Pair("X-Real-IP", "\$remote_addr")),
            proxyAddHeaders = mapOf(Pair("X-User-Agent", "Tunnel"))
        )

    }

    @After
    fun shutDown() {
        client.destroy()
    }
}